stan/math/prim/err/is_mat_finite.hpp

```diff
@@ -11,14 +12,13 @@ namespace math {
  * Return <code>true</code> is the specified matrix is finite.
  * @tparam T Scalar type of the matrix, requires class method
  * <code>.allFinite()</code>
- * @tparam R number of rows or Eigen::Dynamic
- * @tparam C number of columns or Eigen::Dynamic
+ * @tparam EigMat A type derived from `EigenBase`
  * @param y Matrix to test
  * @return <code>true</code> if the matrix is finite
  **/
-template <typename T, int R, int C>
-inline bool is_mat_finite(const Eigen::Matrix<T, R, C>& y) {
-  return y.allFinite();
+template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
+inline bool is_mat_finite(const EigMat& y) {
+  return to_ref(y).allFinite();
```

Comment:
Really odd, if i don't put a `to_ref` here the expression test fails?

Reply:
Hmm, I looked over Eigen source and it is implemented as:
```
#if EIGEN_COMP_MSVC || (defined __FAST_MATH__)
  return derived().array().isFinite().all();
#else
  return !((derived()-derived()).hasNaN());
#endif
```
Some more info here: https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1008

I think we should call `y.array().isFinite().all()` directly to avoid multiple evaluations. It seems that in Eigen they rely on the compiler to optimize dual evaluations. That might work with primitive scalars, but operations on vars have side effects, so I'm pretty sure they would not be optimized.

Reply:
Reading over that thread I'm not sure if the array method actually stops an evaluation. But I just put it up so we'll find out